File: Level4/Problem1/Answer.py
"""
For this final solution we are going to follow a third party solution available online. First we are going to find all possible selections of the number of bunnies possible. For instance if we have a total of 3 bunnies-> [0,1,2] then we first find out all possible combinations of the selection of these bunnies. This includes, [1,2,3],[1],[2],[3],[1,2],[2,3],[1,3]. 

Subset preparation(basically selection of bunnies in all possible ways):
The selection is done in a rather clever way. Firstly observe that the number of ways of all possible selection of bunnies = nCn + nC(n-1) + nC(n-2) + ... + nC1 = 2**n - 1, where n is the total number of bunnies. One important idea here is that the possible selections can be found out using the binary set of numbers from 1 to 2**n - 1. Now if we observe the positions of 1 in all these numbers then we can actually find the all possible selections of these numbers. Its an interesting, consequence of the fact that we have 2**n-1 numbers in total.

Now a bit of graph theory:
What follows is an algorithm similar to the Floyd–Warshall algorithm. What we are doing here is the following.
min=[[]]
for i in range(len):
    for src in range(len):
        if i==0:
            for p in range(len):
                min[src][p] = MAXVALUE # MAXVALUE is a maximum value possible 
            min[src][src]=0
        
        for a in range(len):
            for b in range(len):
                if min[src][a]!=MAXVALUE and min[src][a] + times[a][b] < min[src][b]:
                    min[src][b] = min[src][a] + times[a][b]

The above block of code basically makes sure that from a certain position in the graph if we are travelling to another position in the same row, then it must be having the least time.

Next we find out all possible permutations of the selections that e just did. Then we apply the permutations and find out in which case we get the best results.

"""
import functools
permutations=[]

def comp(a, b):
    # a, b are 2 lists in a list of lists
    if len(a)!=len(b):
        return len(b)-len(a)
    else:# lengths are equal
        return sum(a)-sum(b)

def getSelections(numberOfBunnies):
    numSelections = 2**(numberOfBunnies)-1
    maxLength=len("{0:b}".format(numSelections))
    numbers=[]
    for i in range(1, numSelections+1):
        a=list("{0:b}".format(i))
        if len(a) != maxLength:
            for i in range(len(a), maxLength):
                a=['0']+a
        numbers.append(a)
    # now we have got all the numbers
    # print('numbers: {}'.format(numbers))
    selections=[]
    s=[]
    for i in numbers:
        s=[]
        for j in range(len(i)):
            if i[j]=='1':
                s.append(j)
        
        selections.append(s)
    # print('selections now: {}'.format(selections))
    # [[0, 1, 2], [0, 1], [0, 2], [1, 2], [0], [1], [2], []]
    selections=sorted(selections, key=functools.cmp_to_key((comp)))
    # selections.reverse()
        

    # for i in range(len(selections)): # simple selection sort here because we dp not need to fancy sorting simply because the max number possible is 2**5-1=31
    #     maxVal=len(selections[i])
    #     pos=i
    #     for j in range(i+1, len(selections)):
    #         if len(selections[j])>maxVal:
    #             maxVal=len(selections[j])
    #             pos=j

    #     temp=selections[i][:]
    #     selections[i]=selections[pos][:]
    #     selections[pos]=temp[:]
    
    # print(selections)
    return selections
    
def permute(a, l, r):
    if l==r:
        # print (str(a))
        permutations.append(a[:])
        # yield a
    else:
        for i in range(l,r+1):
            a[l], a[i] = a[i], a[l]
            permute(a, l+1, r)
            a[l], a[i] = a[i], a[l] # backtrack        

def answer(times, time_limit):
    MAXVALUE=10000000
    minMatrix=[]
    #initialization of matrix
    for x in range(len(times)):
        minMatrix.append([])
        for y in range(len(times)):
            minMatrix[x].append(0)


    # print(minMatrix)

    for i in range(len(times)):
        for src in range(len(times)):
            if i==0:
                for p in range(len(times)):
                    minMatrix[src][p] =  MAXVALUE
                minMatrix[src][src]=0
            
            for a in range(len(times)):
                for b in range(len(times)):
                    if minMatrix[src][a]!=MAXVALUE and minMatrix[src][a] + times[a][b] < minMatrix[src][b]:
                        minMatrix[src][b] = minMatrix[src][a] + times[a][b]

    # print('minMatrix: {}'.format(minMatrix))
    # 0,2,1,1,-1,
    # 8,0,1,1,-1,
    # 8,2,0,1,-1,
    # 8,2,1,0,-1,
    # 9,3,2,2,0,

    for src in range(len(times)):
        for i in range(len(times)):
            for j in range(len(times)):
                if minMatrix[src][i] + times[i][j] < minMatrix[src][j]:
                    result=[]
                    for p in range(len(times)-2):
                        result.append(p)
                    return result

    selections=getSelections(len(times)-2)
    # print('permutations')

    for i in range(len(selections)):
        permute(selections[i][:], 0, len(selections[i])-1)
        # print(permutations)
        

        for perm in permutations:
            fromPosition = 0
            time = time_limit
            # print('perm now: {}'.format(perm))
            for val in perm:
                time -= minMatrix[fromPosition][val + 1]
                fromPosition = val + 1

            time -= minMatrix[fromPosition][len(times)-1]
            if time>=0:
                print('selection now: {}'.format(selections[i]))
                del permutations[:]
                return selections[i]

            

        del permutations[:]

    return [] # if nothing can be saved
        # print('---------------------')

        
            
        
        # for j in range(len(permutations)):
        #     f=0
        #     time=time_limit



    
# getSelections(3)
answer( [[0, 2, 2, 2, -1], [9, 0, 2, 2, -1], [9, 3, 0, 2, -1], [9, 3, 2, 0, -1], [9, 3, 2, 2, 0]], 1)

answer( [[0, 1, 1, 1, 1], [1, 0, 1, 1, 1], [1, 1, 0, 1, 1], [1, 1, 1, 0, 1], [1, 1, 1, 1, 0]], 3)

    

                
        

File: Level4/Problem1/test/Answer.java
/*
Running with Bunnies
====================
You and your rescued bunny prisoners need to get out of this collapsing death trap of a space station - and fast! Unfortunately, some of the bunnies have been weakened by their long imprisonment and can't run very fast. Their friends are trying to help them, but this escape would go a lot faster if you also pitched in. The defensive bulkhead doors have begun to close, and if you don't make it through in time, you'll be trapped! You need to grab as many bunnies as you can and get through the bulkheads before they close. 
The time it takes to move from your starting point to all of the bunnies and to the bulkhead will be given to you in a square matrix of integers. Each row will tell you the time it takes to get to the start, first bunny, second bunny, ..., last bunny, and the bulkhead in that order. The order of the rows follows the same pattern (start, each bunny, bulkhead). The bunnies can jump into your arms, so picking them up is instantaneous, and arriving at the bulkhead at the same time as it seals still allows for a successful, if dramatic, escape. (Don't worry, any bunnies you don't pick up will be able to escape with you since they no longer have to carry the ones you did pick up.) You can revisit different spots if you wish, and moving to the bulkhead doesn't mean you have to immediately leave - you can move to and from the bulkhead to pick up additional bunnies if time permits.
In addition to spending time traveling between bunnies, some paths interact with the space station's security checkpoints and add time back to the clock. Adding time to the clock will delay the closing of the bulkhead doors, and if the time goes back up to 0 or a positive number after the doors have already closed, it triggers the bulkhead to reopen. Therefore, it might be possible to walk in a circle and keep gaining time: that is, each time a path is traversed, the same amount of time is used or added.
Write a function of the form answer(times, time_limit) to calculate the most bunnies you can pick up and which bunnies they are, while still escaping through the bulkhead before the doors close for good. If there are multiple sets of bunnies of the same size, return the set of bunnies with the lowest prisoner IDs (as indexes) in sorted order. The bunnies are represented as a sorted list by prisoner ID, with the first bunny being 0. There are at most 5 bunnies, and time_limit is a non-negative integer that is at most 999.
For instance, in the case of
[
  [0, 2, 2, 2, -1],  # 0 = Start
  [9, 0, 2, 2, -1],  # 1 = Bunny 0
  [9, 3, 0, 2, -1],  # 2 = Bunny 1
  [9, 3, 2, 0, -1],  # 3 = Bunny 2
  [9, 3, 2, 2,  0],  # 4 = Bulkhead
]
and a time limit of 1, the five inner array rows designate the starting point, bunny 0, bunny 1, bunny 2, and the bulkhead door exit respectively. You could take the path:
Start End Delta Time Status
    -   0     -    1 Bulkhead initially open
    0   4    -1    2
    4   2     2    0
    2   4    -1    1
    4   3     2   -1 Bulkhead closes
    3   4    -1    0 Bulkhead reopens; you and the bunnies exit
With this solution, you would pick up bunnies 1 and 2. This is the best combination for this space station hallway, so the answer is [1, 2].
Languages
=========
To provide a Python solution, edit solution.py
To provide a Java solution, edit solution.java
Test cases
==========
Inputs:
    (int) times = [[0, 1, 1, 1, 1], [1, 0, 1, 1, 1], [1, 1, 0, 1, 1], [1, 1, 1, 0, 1], [1, 1, 1, 1, 0]]
    (int) time_limit = 3
Output:
    (int list) [0, 1]
Inputs:
    (int) times = [[0, 2, 2, 2, -1], [9, 0, 2, 2, -1], [9, 3, 0, 2, -1], [9, 3, 2, 0, -1], [9, 3, 2, 2, 0]]
    (int) time_limit = 1
Output:
    (int list) [1, 2]
*/
package com.google.challenges; 
import java.util.*;
import java.io.*;

class subsetComparator implements Comparator<LinkedList<Integer>> {
    // @Override
    public int compare(LinkedList<Integer> a, LinkedList<Integer> b) {

    	if (a.size() != b.size())
    		return (b.size() - a.size());
    	else {
    		for (int i = 0; i < a.size(); i++) {

    			if (a.get(i) < b.get(i))
    				return a.get(i) - b.get(i);

    			else if (a.get(i) > b.get(i))
    				return a.get(i) - b.get(i);

    		}
    		return 1;
    	}
    }
}

public class Answer {
    
	static Set<LinkedList<Integer>> set = new HashSet<LinkedList<Integer>>();
	static LinkedList<LinkedList<Integer>> permutations = new LinkedList<LinkedList<Integer>>();

	public static void permute(LinkedList<LinkedList<Integer>> subsets, int at, int index) {

		LinkedList<Integer> subset = subsets.get(index);

		for (int i = at; i < subset.size(); i++) {

			Collections.swap(subset, i, at);

			if (!set.contains(subset)) {

				LinkedList<Integer> new_perm = new LinkedList<Integer> (subset);

				set.add(subset);
				permutations.add(new_perm);

			}

			permute(subsets, at+1, index);
			
			Collections.swap(subset, at, i);
		}


	}


	public static LinkedList<LinkedList<Integer>> subsets(int bunnies) {
		//bunnies here is the number of bunnies.
		LinkedList<LinkedList<Integer>> subsets = new LinkedList<LinkedList<Integer>>();

		int max = (int)Math.pow(2,bunnies) - 1;
		//for a value of bunnies=3, like in the first test case the value of max is going to be 7 which is the highest number we can go in binary in 3 digits.

		int maxLen = Integer.toBinaryString(max).length();

		String bin;

		for (; max >= 0 ; max--) {
			
			bin = Integer.toBinaryString(max);

			int diff = maxLen - bin.length();

			for (int l = 0; l < diff; l++)
				bin = "0" + bin;

			LinkedList<Integer> subset = new LinkedList<Integer>();

			for (int i = 0 ; i < bin.length(); i++) {
				if (bin.charAt(i) == '1') {
					subset.add(i);
				}
			}

			subsets.add(subset);
		}

		Collections.sort(subsets, new subsetComparator());
// 		System.out.println(subsets);
		return subsets;

	}

	
    public static int[] answer(int[][] times, int time_limit) { 

        // Your code goes here.
		int len = times.length;

		LinkedList<LinkedList<Integer>> subsets = subsets(len-2);
		/*
		A rather intriguing thing happens in the previous line. What is done is that we try and get all possible subsets of a the set [0,1,2] where 0 is the first bunny, 1 is the second bunny and 2 is the third bunny. The way we get all the subsets for the set [0,1,2] is the following:
		Its like the total number of ways we can select 3 bunnies.
		Now observe that in binary numbers for each numbers until we reach the number 2*n-1 where n is the number of values among which we are selecting, we ultimately end up getting all the possible subsets. Its a rather clever way of getting all possible subsets of a number. 
		*/
		int[][] min = new int[len][len];
		
		int src = 0;
		
	    for (int z = 0; z < len; z++)
	    {
    		for (src = 0; src < len; src++)
    		{
    		    if (z==0) {
        			for (int i=0; i<len; ++i)
        				min[src][i] = Integer.MAX_VALUE;
						//Integer.MAX_VALUE is the maximum value that an integer can have which is typically 231-1.
        			min[src][src] = 0;
    		    }
				// System.out.println("currently z="+Integer.toString(z)+": and src =" + Integer.toString(src));
				// for(int a=0;a<len;a++)
				// {
				// 	for(int b=0;b<len;b++)
				// 	{
				// 		System.out.print(Integer.toString(min[a][b])+",");
				// 	}
				// 	System.out.println();
				// }

    			for (int i=0; i<len; ++i)
    	            for (int j=0; j<len; ++j)
    	                if (min[src][i] != Integer.MAX_VALUE && min[src][i] + times[i][j] < min[src][j])
    	                    min[src][j] = min[src][i] + times[i][j];
				/*
				Here basically a relaxation step goes on. Recall that in graph theory in case of the bellman ford algorithm or the djikstra algorithm, a relaxation step is followed. The step goes in the following way: if we have a node X that is connected to a node Y which in turn is connected to a node Z and if node X is also connected directly to node Z then the relaxation step says that if w(X,Z)<w(X,Y)+w(Y,Z) then w(X,Z)=w(X,Y)+w(Y,Z). This is a fundamental step in the process of finding the shortest path from one node (source) to all other nodes. It is also called single source shortest path algorithm. 

				Here what we are doing is very clever to say the least. Its an interesting step. What we are actually doing here is that we are taking the first node in the first row initially, then we are checking the distance of that node with any node in the whole graph adding that distance and checking whether that distance is more than the distance stored for going from that other node (i,j) to src,j.			
				*/
    	    }
// 			System.out.println("-------------------------------------------");
// 			for (int a=0;a< len;a++)
// 			{
// 				for(int b=0;b<len;b++)
// 				{
// 					System.out.print(Integer.toString(min[a][b])+',');
// 				}
// 				System.out.println();
// 			}
	    }
// 		System.out.println("-------------------------------------------");
// 		System.out.println("-------------------------------------------");
// 	   for (int i=0;i< len;i++)
// 	   {
// 		   for(int j=0;j<len;j++)
// 		   {
// 			   System.out.print(Integer.toString(min[i][j])+',');
// 		   }
// 		   System.out.println();
// 	   }
	   
		for (src = 0; src < len; src++)
		{
			for (int i=0; i<len; ++i)
	        {
	            for (int j=0; j<len; ++j)
	            {
	                if (min[src][i] + times[i][j] < min[src][j]) {
	                    int[] cycle = new int[len-2];
	                    
	                    for (int c = 0 ; c < len-2 ; c++)
	                        cycle[c] = c;
				// 		System.out.println("cycle found");
	                    return cycle;
	                }
	            }
	        }
	    }
	    
	    // -=-
	    
	    int from;
	    int time;

	    for (int index = 0; index < subsets.size(); index++)
	    {
			permute(subsets, 0, index);
			
			//Collections.sort(permutations, new subsetComparator());
// 			System.out.println("permutations");
// 			System.out.println(permutations);
			
			for (int p = 0 ; p < permutations.size(); p++ ) {
			    
				from = 0;
				time = time_limit;

				LinkedList<Integer> perm = permutations.get(p);
				// System.out.println("perm now:" + perm);
				// for (int i=0;i<perm.size();i++)
				// {
				// 	System.out.print(perm.get(i)+",");
				// }
				for (int i = 0; i < perm.size(); i++) {

					time -= min[from][perm.get(i)+1];
					from = perm.get(i)+1;
					
				}

				time -= min[from][len-1];

				if (time >= 0) {
				    
				    int[] result = new int[subsets.get(index).size()];
				    
				    for (int r = 0 ; r < subsets.get(index).size(); r++)
				        result[r] = subsets.get(index).get(r);
				    
				    set.clear();
			        permutations.clear();
			    
				    return result;
				}

			}

			set.clear();
			permutations.clear();

		}

		int[] result = new int[0];

		return result;
    } 
// 	public static void main(String args[])throws IOException
// 	{
// 		// int times[][] =new int [][]{{0, 2, 2, 2, -1}, {9, 0, 2, 2, -1}, {9, 3, 0, 2, -1}, {9, 3, 2, 0, -1}, {9, 3, 2, 2, 0}};
// 		int times[][]=new int[][]{{0,1,1,1,1},{1,0,1,1,1},{1,1,0,1,1},{1,1,1,0,1},{1,1,1,1,0}};
// 		int time_limit=3;
// 		// int time_limit=1;
// 		// Answer obj=new Answer();
// 		// System.out.println(obj.answer(times, time_limit));
// 		int[] a=answer(times, time_limit);
// 		for (int i=0;i<a.length;i++){
// 			System.out.println(a[i]+",");
// 		}

// 	}

}